# Zomato Bangalore Dataset Analysis & Visualization

## 4.Data Cleaning & Pre-processing

## 🔹 Importing Required Libraries
We begin by importing the essential Python libraries for data analysis and cleaning. These include:
- `pandas` for data manipulation,
- `numpy` for numerical operations,


In [4]:
import numpy as np 
import pandas as pd


### 🔹Load Dataset

The dataset is loaded using `pd.read_csv()` and displayed to understand its structure, dimensions, and first few records.


In [5]:
df = pd.read_csv(r"C:\Users\abhir\Downloads\zomato.csv")
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

### 🔹 4.1 Drop irrelevant columns that do not contribute to analysis

These columns do not contribute to our analysis. Dropping reduces dataset size and complexity.

In [3]:
columns_to_drop = ['url', 'address', 'phone', 'reviews_list', 'menu_item', 'listed_in(city)']
df.drop(columns= columns_to_drop, inplace = True)

### 🔹4.2 Remove duplicate rows to avoid skewed results 

Duplicates may skew data and introduce bias. Removing ensures each restaurant is unique.

In [4]:
df.drop_duplicates(inplace= True)

### 🔹 4.3 Drop Rows with Missing Values in Critical Columns

Missing values can cause problems in calculations and model training. For critical columns (rate, approx_cost(for two people), location), missing entries are removed

In [5]:
df.dropna(subset= ['rate', 'approx_cost(for two people)', 'location'], inplace = True)

### 🔹 4.4 Fill Missing Values in Less Critical Columns

For non-critical columns (cuisines, rest_type, dish_liked), missing values are replaced with default placeholders.

In [6]:
df['cuisines'] = df['cuisines'].fillna('Unknown')
df['rest_type'] = df['rest_type'].fillna('Others')
df['dish_liked'] = df['dish_liked'].fillna('Not specified')

### 🔹 4.5 Clean and Convert rate Column

The `rate` column contains strings like `'4.1/5'` or `'-'`.  
Steps:
1. Remove the `/5` part.
2. Strip spaces.
3. Replace `'-'` with NaN (since it's not a valid rating).
4. Convert to `float` type for calculations.


In [7]:
df['rate'] = df['rate'].astype(str).str.replace('/5', '', regex=False).str.strip()
df['rate'] = df['rate'].replace('-', np.nan)
df['rate'] = pd.to_numeric(df['rate'], errors='coerce')

### 🔹 4.6 Clean and Rename approx_cost(for two people)

Removes commas, converts to float, and renames to cost.

In [8]:
df['approx_cost(for two people)']= df['approx_cost(for two people)'].astype(str).str.replace(",", "").astype(float)
df.rename(columns= {'approx_cost(for two people)':'cost'}, inplace = True)

### 🔹4.7 Add new column with cost per person

Dividing `cost` by 2 provides an individual meal price estimate.

In [9]:
df['cost_per_person'] = df['cost'] / 2

### 🔹4.8 Extract the first listed cuisine as the primary one

Extracting the first cuisine listed gives us the main food category each restaurant is known for.

In [10]:
df['primary_cuisines'] = df['cuisines'].str.split(',').str[0]

### 🔹4.9 Convert 'Yes/No' to boolean in online order and table booking

For `online_order` and `book_table`,  
we map `'Yes'` → `True` and `'No'` → `False` to facilitate logical/boolean filtering.


In [11]:
df['online_order'] = df['online_order'].map({'Yes': True, 'No': False})
df['book_table'] = df['book_table'].map({'Yes': True, 'No': False})


### 🔹4.10 Normalize column names: lowercase and underscores instead of spaces

Making all column names lowercase and separating words with underscores:
- Improves typing convenience
- Avoids case-sensitivity issues


In [12]:
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

### 🔹4.11 Remove rows with zero or negative cost values

Any restaurant entry showing `0` or negative `cost` values is unrealistic and likely due to data entry errors.  
Removing them ensures more accurate cost-based analysis.


In [13]:
df[df['cost'] > 0 ]

url  \
0      https://www.zomato.com/bangalore/jalsa-banasha...   
1      https://www.zomato.com/bangalore/spice-elephan...   
2      https://www.zomato.com/SanchurroBangalore?cont...   
3      https://www.zomato.com/bangalore/addhuri-udupi...   
4      https://www.zomato.com/bangalore/grand-village...   
...                                                  ...   
51712  https://www.zomato.com/bangalore/best-brews-fo...   
51713  https://www.zomato.com/bangalore/vinod-bar-and...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   
51715  https://www.zomato.com/bangalore/chime-sherato...   
51716  https://www.zomato.com/bangalore/the-nest-the-...   

                                                 address  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...   
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name  online_order  \
0                                                  Jalsa          True   
1                                         Spice Elephant          True   
2                                        San Churro Cafe          True   
3                                  Addhuri Udupi Bhojana         False   
4                                          Grand Village         False   
...                                                  ...           ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...         False   
51713                           Vinod Bar And Restaurant         False   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...         False   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...         False   
51716                       The Nest - The Den Bengaluru         False   

       book_table  rate  votes                             phone  \
0            True   4.1    775    080 42297555\r\n+91 9743772233   
1           False   4.1    787                      080 41714161   
2           False   3.8    918                    +91 9663487993   
3           False   3.7     88                    +91 9620009302   
4           False   3.8    166  +91 8026612447\r\n+91 9901210005   
...           ...   ...    ...                               ...   
51712       False   3.6     27                      080 40301477   
51713       False   NaN      0                    +91 8197675843   
51714       False   NaN      0                               NaN   
51715        True   4.3    236                      080 49652769   
51716       False   3.4     13                    +91 8071117272   

                         location            rest_type  \
0                    Banashankari        Casual Dining   
1                    Banashankari        Casual Dining   
2                    Banashankari  Cafe, Casual Dining   
3                    Banashankari          Quick Bites   
4                    Basavanagudi        Casual Dining   
...                           ...                  ...   
51712                  Whitefield                  Bar   
51713                  Whitefield                  Bar   
51714                  Whitefield                  Bar   
51715  ITPL Main Road, Whitefield                  Bar   
51716  ITPL Main Road, Whitefield   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, 

### 🔹4.12 Remove unrealistic rating values (e.g., greater than 5)

This step filters the dataset to keep only rows where the rating is **5 or below**, removing any invalid or unrealistic ratings.


In [14]:
df = df[df['rate'] <= 5]

### 🔹4.13 Remove rows where votes are missing or non-

This step keeps only the rows where the **votes** value is greater than zero, removing entries with missing or non-positive vote counts.


In [15]:
df[df['votes'] > 0]

url  \
0      https://www.zomato.com/bangalore/jalsa-banasha...   
1      https://www.zomato.com/bangalore/spice-elephan...   
2      https://www.zomato.com/SanchurroBangalore?cont...   
3      https://www.zomato.com/bangalore/addhuri-udupi...   
4      https://www.zomato.com/bangalore/grand-village...   
...                                                  ...   
51709  https://www.zomato.com/bangalore/the-farm-hous...   
51711  https://www.zomato.com/bangalore/bhagini-2-whi...   
51712  https://www.zomato.com/bangalore/best-brews-fo...   
51715  https://www.zomato.com/bangalore/chime-sherato...   
51716  https://www.zomato.com/bangalore/the-nest-the-...   

                                                 address  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...   
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51709  136, SAP Labs India, KIADB Export Promotion In...   
51711  139/C1, Next To GR Tech Park, Pattandur Agraha...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name  online_order  \
0                                                  Jalsa          True   
1                                         Spice Elephant          True   
2                                        San Churro Cafe          True   
3                                  Addhuri Udupi Bhojana         False   
4                                          Grand Village         False   
...                                                  ...           ...   
51709                         The Farm House Bar n Grill         False   
51711                                            Bhagini         False   
51712  Best Brews - Four Points by Sheraton Bengaluru...         False   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...         False   
51716                       The Nest - The Den Bengaluru         False   

       book_table  rate  votes                             phone  \
0            True   4.1    775    080 42297555\r\n+91 9743772233   
1           False   4.1    787                      080 41714161   
2           False   3.8    918                    +91 9663487993   
3           False   3.7     88                    +91 9620009302   
4           False   3.8    166  +91 8026612447\r\n+91 9901210005   
...           ...   ...    ...                               ...   
51709       False   3.7     34    +91 9980121279\n+91 9900240646   
51711       False   2.5     81                      080 65951222   
51712       False   3.6     27                      080 40301477   
51715        True   4.3    236                      080 49652769   
51716       False   3.4     13                    +91 8071117272   

                         location            rest_type  \
0                    Banashankari        Casual Dining   
1                    Banashankari        Casual Dining   
2                    Banashankari  Cafe, Casual Dining   
3                    Banashankari          Quick Bites   
4                    Basavanagudi        Casual Dining   
...                           ...                  ...   
51709                  Whitefield   Casual Dining, Bar   
51711                  Whitefield   Casual Dining, Bar   
51712                  Whitefield                  Bar   
51715  ITPL Main Road, Whitefield                  Bar   
51716  ITPL Main Road, Whitefield   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, 

### 🔹4.14 Filter Sparse Locations

Removed locations with fewer than 10 listings to ensure reliable location-based insights.


In [16]:
location_counts = df['location'].value_counts()
valid_locations = location_counts[location_counts >= 10].index
df = df[df['location'].isin(valid_locations)]

### 🔹4.15 Preview of first five records of cleaned data

In [17]:
df.head(5)

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

   online_order  book_table  rate  votes                             phone  \
0          True        True   4.1    775    080 42297555\r\n+91 9743772233   
1          True       False   4.1    787                      080 41714161   
2          True       False   3.8    918                    +91 9663487993   
3         False       False   3.7     88                    +91 9620009302   
4         False       False   3.8    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines   cost  \
0  North Indian, Mughlai, Chinese  800.0   
1     Chinese, North Indian, Thai  800.0   
2          Cafe, Mexican, Italian  800.0   
3      South Indian, North Indian  300.0   
4        North Indian, Rajasthani  600.0   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  cost_per_person primary_cuisines  
0          Buffet    Banashankari            400.0     North Indian  
1          Buffet    Banashankari            400.0          Chinese  
2          Buffet    Banashankari            400.0             Cafe  
3          Buffet    Banashankari            150.0     South Indian  
4          Buffet    Banashankari            300.0     North Indian

### 🔹Exporting the Cleaned Data to CSV

In [19]:
df.to_csv('cleaned_zomato_dataset', index= False)

## 🔹Interpretation Summary

By the end of cleaning, we:
- Eliminated irrelevant & duplicate data
- Handled missing values wisely
- Standardized formats and data types
- Added useful new columns
- Kept only valid, high-quality rows

The dataset is now **analysis-ready**.
